File: Tictactoe/board.hs
module Tictactoe.Board where

import Common

newtype Board = Board
  { tiles :: [(Int, Player)]
  }

empty = Board{tiles = []}

add board tile = Board{tiles = tile : tiles board}

elems indices b = map (`lookup` tiles b) indices

rows = allLines row where row n = elems [(n * 3) .. (n * 3) + 2]

cols = allLines col where col n = elems [n, n + 3, n + 6]

(dia1, dia2) = (elems [0, 4, 8], elems [2, 4, 6])

allLines line board = map (`line` board) [0 .. 2]

instance Show Board where
  show = unlines . map (unwords . map showPlayer) . rows

File: tictactoe/board.py
Board = tuple[int, int]

# Configurations for a winning line
wins = [
    7, 56, 448, # row
    73, 146, 292, # col
    273, 84 # diag
]

def exists(board: Board, tile: int) -> bool:
    os, xs = board

    return bool(os & (1 << tile)) or bool(xs & (1 << tile))

def at(board: Board, coord: int):
    os, xs = board

    if bool(os & (1 << coord)): return 'O'
    if bool(xs & (1 << coord)): return 'X'
    return '_'

def turn(board: Board) -> int:
    os, xs = board
    return ((os | xs).bit_count() + 1) % 2

def finished_player(player: int) -> bool:
    return any((player & win) == win for win in wins)
